hw/misc/hermes.c

```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 {
 }
 
+static int __do_dma(struct HermesState *hermes, hwaddr desc_addr,
+                       uint8_t num_desc, bool h2c)
+{
+    struct hermes_dma_desc *desc;
+    hwaddr src_addr, dst_addr, nxt_addr;
+    uint8_t nxt_adj;
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma_num_adj(num_desc - 1);
+    desc = g_new(struct hermes_dma_desc, num_desc);
+    if (!desc) {
+        fprintf(stderr, "[Hermes] Failed to alloc memory for DMA descriptor\n");
+        return -1;
+    }
+
+    /* Read DMA descriptors */
+    pci_dma_read(&hermes->pdev, desc_addr, desc, num_desc * sizeof(*desc));
+    for (int i = 0; i < num_desc; i++) {
+        trace_hermes_dma_desc(desc[i].ctrl, desc[i].len, desc[i].src_addr_lo,
+                              desc[i].src_addr_hi, desc[i].dst_addr_lo,
+                              desc[i].dst_addr_hi, desc[i].nxt_addr_lo,
+                              desc[i].nxt_addr_hi);
+    }
+
+    /* DMA all descriptors in this block */
+    for (int i = 0; i < num_desc; i++) {
+        src_addr = ((uint64_t) desc[i].src_addr_hi << 32) | desc[i].src_addr_lo;
+        dst_addr = ((uint64_t) desc[i].dst_addr_hi << 32) | desc[i].dst_addr_lo;
+
+        if (h2c) {
+            trace_hermes_dma("H2C", src_addr, dst_addr);
+            pci_dma_read(&hermes->pdev, src_addr, bar4_base + dst_addr,
+                         desc[i].len);
+        } else {
+            trace_hermes_dma("C2H", src_addr, dst_addr);
+            pci_dma_write(&hermes->pdev, dst_addr, bar4_base + src_addr,
+                          desc[i].len);
+        }
+    }
+
+    nxt_addr = ((uint64_t) desc[num_desc - 1].nxt_addr_hi << 32)
+               | desc[num_desc - 1].nxt_addr_lo;
+    if (nxt_addr) {
+        nxt_adj = (desc[num_desc - 1].ctrl >> 8) & 0x3F;
+        num_desc += __do_dma(hermes, nxt_addr, nxt_adj + 1, h2c);
+    }
+
+    g_free(desc);
+
+    return num_desc;
+}
+
+static void do_dma(HermesState *hermes, bool h2c)
+{
+    struct hermes_bar2_engine_reg *engine_reg;
+    struct hermes_bar2_sgdma_reg *sgdma_reg;
+    struct hermes_bar2 *bar2 = &hermes->bar2;
+    struct hermes_bar2_irq_reg *irq = &bar2->irq;
+    hwaddr desc_addr;
+    unsigned irq_vector;
+    unsigned num_desc;
+
+    if (h2c) {
```

Comment:
Still don't like these h2c bools... Better to restructure your data a bit to make this cleaner...

This very rough, certainly won't compile mockup should hopefully demonstrate what I'm thinking:

```diff
diff --git a/hw/misc/hermes.c b/hw/misc/hermes.c
index 6c65c458c0a9..aa125b6c2b57 100644
--- a/hw/misc/hermes.c
+++ b/hw/misc/hermes.c
@@ -83,7 +83,7 @@ typedef struct HermesState HermesState;
  * https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf
  * (note that in this document it uses BAR1, but here and on AWS we use BAR2)
  */
-struct hermes_bar2_engine_reg {
+struct hermes_bar2_channel_reg {
     uint32_t identifier;         /* 0x00 */
     uint32_t control;            /* 0x04, 0x08 and 0x0C */
     uint32_t status;             /* 0x40 and 0x44 */
@@ -153,13 +153,17 @@ struct hermes_bar2_msix_pba_reg {
     uint32_t pba;                   /* 0xFE0 */
 };
 
+struct hermes_bar2_dir_reg {
+    struct hermes_bar2_engine_reg channel;
+    struct hermes_bar2_sgdma_reg sgdma;
+    void (*perform_dma)(HermesState hermes, hwaddr src_addr, dst_addr);
+    int irq_vector;
+};
+
 struct hermes_bar2 {
-    struct hermes_bar2_engine_reg h2c;
-    struct hermes_bar2_engine_reg c2h;
+    struct hermes_bar2_dir_reg h2c, c2h;
     struct hermes_bar2_irq_reg irq;
     struct hermes_bar2_cfg_reg cfg;
-    struct hermes_bar2_sgdma_reg h2c_sgdma;
-    struct hermes_bar2_sgdma_reg c2h_sgdma;
     struct hermes_bar2_sgdma_common_reg sgdma_common;
     struct hermes_bar2_msix_pba_reg msix_pba;
 } bar2_init = {
@@ -274,110 +278,109 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 {
 }
 
-static int __do_dma(struct HermesState *hermes, hwaddr desc_addr,
-                       uint8_t num_desc, bool h2c)
+static void hermes_h2c_dma(HermesState hermes, hwaddr src_addr, dst_addr)
+{
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma("H2C", src_addr, dst_addr);
+    pci_dma_read(&hermes->pdev, src_addr, bar4_base + dst_addr,
+                 desc[i].len);
+
+}
+
+static void hermes_c2h_dma(HermesState hermes, hwaddr src_addr, dst_addr)
+{
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma("C2H", src_addr, dst_addr);
+    pci_dma_write(&hermes->pdev, dst_addr, bar4_base + src_addr,
+                  desc[i].len);
+}
+
+static int hermes_execute_descs(struct HermesState *hermes,
+		struct hermes_bar2_dir_reg *dir, hwaddr *desc_addr, uint8_t *num_desc)
 {
     struct hermes_dma_desc *desc;
     hwaddr src_addr, dst_addr, nxt_addr;
     uint8_t nxt_adj;
-    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
 
-    trace_hermes_dma_num_adj(num_desc - 1);
-    desc = g_new(struct hermes_dma_desc, num_desc);
+    trace_hermes_dma_num_adj(*num_desc - 1);
+    desc = g_new(struct hermes_dma_desc, *num_desc);
     if (!desc) {
         fprintf(stderr, "[Hermes] Failed to alloc memory for DMA descriptor\n");
         return -1;
     }
 
     /* Read DMA descriptors */
-    pci_dma_read(&hermes->pdev, desc_addr, desc, num_desc * sizeof(*desc));
-    for (int i = 0; i < num_desc; i++) {
+    pci_dma_read(&hermes->pdev, *desc_addr, desc, *num_desc * sizeof(*desc));
+    for (int i = 0; i < *num_desc; i++) {
         trace_hermes_dma_desc(desc[i].ctrl, desc[i].len, desc[i].src_addr_lo,
                               desc[i].src_addr_hi, desc[i].dst_addr_lo,
                               desc[i].dst_addr_hi, desc[i].nxt_addr_lo,
                               desc[i].nxt_addr_hi);
-    }
 
-    /* DMA all descriptors in this block */
-    for (int i = 0; i < num_desc; i++) {
         src_addr = ((uint64_t) desc[i].src_addr_hi << 32) | desc[i].src_addr_lo;
         dst_addr = ((uint64_t) desc[i].dst_addr_hi << 32) | desc[i].dst_addr_lo;
 
-        if (h2c) {
-            trace_hermes_dma("H2C", src_addr, dst_addr);
-            pci_dma_read(&hermes->pdev, src_addr, bar4_base + dst_addr,
-                         desc[i].len);
-        } else {
-            trace_hermes_dma("C2H", src_addr, dst_addr);
-            pci_dma_write(&hermes->pdev, dst_addr, bar4_base + src_addr,
-                          desc[i].len);
-        }
+        perform_dma(hermes, src_addr, dst_addr);
     }
 
-    nxt_addr = ((uint64_t) desc[num_desc - 1].nxt_addr_hi << 32)
-               | desc[num_desc - 1].nxt_addr_lo;
-    if (nxt_addr) {
-        nxt_adj = (desc[num_desc - 1].ctrl >> 8) & 0x3F;
-        num_desc += __do_dma(hermes, nxt_addr, nxt_adj + 1, h2c);
-    }
+    *desc_addr = ((uint64_t) desc[i - 1].nxt_addr_hi << 32)
+                 | desc[i - 1].nxt_addr_lo;
+    *num_desc = (desc[i - 1].ctrl >> 8) & 0x3F
 
     g_free(desc);
 
-    return num_desc;
+    return i;
 }
 
-static void do_dma(HermesState *hermes, bool h2c)
+static void hermes_do_dma(HermesState *hermes, struct hermes_bar2_dir_reg *dir)
 {
-    struct hermes_bar2_engine_reg *engine_reg;
-    struct hermes_bar2_sgdma_reg *sgdma_reg;
     struct hermes_bar2 *bar2 = &hermes->bar2;
     struct hermes_bar2_irq_reg *irq = &bar2->irq;
     hwaddr desc_addr;
     unsigned irq_vector;
     unsigned num_desc;
-
-    if (h2c) {
-        sgdma_reg = &bar2->h2c_sgdma;
-        engine_reg = &bar2->h2c;
-        irq_vector = 0;
-    } else {
-        sgdma_reg = &bar2->c2h_sgdma;
-        engine_reg = &bar2->c2h;
-        irq_vector = 1;
-    }
+    int ret;
 
     /* Reset number of completed descriptors */
-    engine_reg->cmp_desc_count = 0;
+    dir->channel.cmp_desc_count = 0;
 
     /* Set engine as busy */
-    atomic_or(&engine_reg->status, 0x1);
+    atomic_or(&dir->channel.status, 0x1);
 
     /*
      * There is always at least one descriptor, plus the adjacent ones (which
      * could be 0). Only bits 5:0 of the register are defined
      */
-    num_desc = 1 + (sgdma_reg->desc_num_adj & 0x3F);
-    desc_addr = ((uint64_t) sgdma_reg->desc_high_addr) << 32 |
-                sgdma_reg->desc_low_addr;
-    num_desc = __do_dma(hermes, desc_addr, num_desc, h2c);
+    num_desc = 1 + (dir->sgdma.desc_num_adj & 0x3F);
+    desc_addr = ((uint64_t) dir->sgdma->desc_high_addr) << 32 |
+        		dir->sgdma->desc_low_addr;
+    engine_reg->cmp_desc_count = 0;
+
+    while (1) {
+        ret = hermes_execute_descs(hermes, dir, &desc_addr, &num_desc);
+        if (ret < 0 || !desc_addr)
+            break;
 
-    /* Set number of completed descriptors */
-    engine_reg->cmp_desc_count = num_desc;
+        /* Set number of completed descriptors */
+        dir->channel.cmp_desc_count += ret;
+    }
 
     /* Set engine as not busy */
-    atomic_and(&engine_reg->status, ~1);
+    atomic_and(&dir->channel.status, ~1);
 
-    if (irq->chnl_inter_enable_mask & (irq_vector + 1)) {
+    if (irq->chnl_inter_enable_mask & (dir->irq_vector + 1)) {
         /* Set interrupt source */
         irq->chnl_inter_request =  irq->chnl_inter_enable_mask &
-                                   (irq_vector + 1);
+                                   (dir->irq_vector + 1);
 
         /*
          * Send interrupt. Since we currently have only one channel for H2C and
          * one for C2H, we have that IRQ 0 is H2C and IRQ 1 is C2H
          */
         trace_hermes_msix_notify(irq_vector);
-        msix_notify(PCI_DEVICE(hermes), irq_vector);
+        msix_notify(PCI_DEVICE(hermes), dir->irq_vector);
     }
 }


```

Reply:
Thanks for the suggestion and code. Fixing it.